File: interface/client/controllers/AccountController.ts
// Copyright © Spatial Corporation. All rights reserved.

import { Account, CreateAccountOptions } from "@sptlco/data";
import { Controller } from "..";

export class AccountController extends Controller {
  /**
   * Create a new account.
   * @param options Configurable options for the account.
   * @returns An account.
   */
  public create = (options: CreateAccountOptions) => this.post<Account>("accounts", options);

  /**
   * Update an account.
   * @param account The account to update.
   * @returns The updated account.
   */
  public update = (account: Account) => this.patch<Account>("accounts", account);

  /**
   * Delete an account.
   * @param id The account to delete.
   * @returns The response from the server.
   */
  public del = (id: string) => this.delete(`accounts/${id}`);
}

File: cloud/Server/API/AccountController.cs
// Copyright © Spatial Corporation. All rights reserved.

using Spatial.Cloud.Data.Accounts;
using Spatial.Cloud.Data.Scopes;
using Spatial.Extensions;
using Spatial.Identity.Authorization;
using Spatial.Persistence;

namespace Spatial.Cloud.API;

/// <summary>
/// A controller for account functions.
/// </summary>
[Path("accounts")]
public class AccountController : Controller
{
    /// <summary>
    /// Create a new <see cref="Account"/>.
    /// </summary>
    /// <param name="options">Configurable options for the <see cref="Account"/>.</param>
    /// <returns>An <see cref="Account"/>.</returns>
    [POST]
    //[Authorize(Scope.Accounts.Create)]
    public async Task<Account> CreateAccountAsync([Body] CreateAccountOptions options)
    {
        if (Resource<Account>.FirstOrDefault(a => a.Email == options.Email) is Account account)
        {
            throw new UserError("The user already exists.");
        }

        account = new Account {
            Name = options.Name,
            Email = options.Email,
            Metadata = options.Metadata
        };

        return account.Store();
    }

    /// <summary>
    /// Update an account.
    /// </summary>
    /// <param name="account">The account to update.</param>
    /// <returns>The updated account.</returns>
    [PATCH]
    //[Authorize(Scope.Accounts.Update)]
    public async Task<Account> UpdateAccountAsync([Body] Account account)
    {
        return account.Save();
    }

    /// <summary>
    /// Delete an account.
    /// </summary>
    /// <param name="account">The account to delete.</param>
    [DELETE]
    [Path("{account}")]
    //[Authorize(Scope.Accounts.Delete)]
    public async Task DeleteAccountAsync(string account)
    {
        Resource<Account>.RemoveOne(a => a.Id == account);
    }
}